RFC.md

```diff
@@ -1,12 +1,295 @@
 # Introducing a new multidimensional array indexing intrinsic
 
-# Intrinsic specification
+## The request for change: A new instruction, `multidim_array_index`.
 
-# Example
-## allocating something of that shape
-## indexing into something of that shape
-## loop analyses that uses this
+We propose the addition of a new instruction, called `multidim_array_index`. 
+This will allow us to represent array indexing into an array `A[d1][d2][..][dk]`
+as `multidim_array_index A d1, d2, d3, ... dk` instead of flattening
+the information into `gep A, d1 * n1 + d2 * n2 + ... + dk * nk`. The former
+unflattened representation is advantageous for analysis and optimisation. It also
+allows us to represent array indexing semantics of languages such as Fortran
+and Chapel, which differs from that of C based array indexing.
+	
+## Motivating the need for a multi-dimensional array indexing intrinsic
+There are primarily one of two views of array indexing involving
+multiple dimensions most languages take, which we discuss to motivate 
+the need for a multi-dimensional array index. This consideration impacts
+the kinds of analysis we can perform on the program. In Polly, we care about
+dependence analysis, so the examples here will focus on that particular problem.
 
-# The math, how to express every conceivable array indexing scheme known to man
+Let us consider an array indexing operation of the form:
+```cpp
+int ex1(int n, int m, B[n][m], int x1, int x2, int y1, int y2) {
+	assert(x1 != x2);
```

Comment:
`_builtin_assume` might be what you want; typically compilers do not optimize based on asserts.

Reply:
Fixed, thanks!